valohai_cli/utils/ssh.py

```diff
@@ -0,0 +1,159 @@
+                            break
+        except OSError:  # e.g. unreadable key
+            pass
+
+
+def select_private_key_from_possible_directories() -> str:
+    default_ssh_dir = os.path.expanduser("~/.ssh")
+    cli_config_dir = get_settings_file_name("")
+    all_keypairs: list[Keypair] = []
+    all_dirs = [cli_config_dir, default_ssh_dir]
+    for dir in all_dirs:
+        if not os.path.isdir(dir):
+            continue
+        all_keypairs.extend(get_ssh_key_pairs_from_dir(dir))
+
+    if not all_keypairs:
+        raise click.UsageError(
+            f"No SSH key pairs found in {', '.join(all_dirs)}. Please try to generate new SSH key pair",
+        )
+
+    keypair_options = [
+        {
+            "name": keypair.private_path,
+            "description": f"Public: {keypair.public_path}",
+        }
+        for keypair in all_keypairs
+    ]
+
+    selected_key = prompt_from_list(prompt="Select the SSH key pair you want to use", options=keypair_options)
+    return str(selected_key["name"])
+
+
+def fetch_status_events(execution: dict, first_n: Optional[int] = None) -> Generator:
+    params = {}
+    if first_n is not None:
+        params["first_n"] = first_n
+    events_response = request("get", f"{execution['url']}status_events/", params=params).json()
+    yield from events_response.get("status_events", [])
+
+
+def _get_execution_from_counter(counter: int) -> dict:
+    project = get_project(require=True)
+    assert project
+
+    execution = project.get_execution_from_counter(
+        counter=counter,
+        params={
+            "exclude": "metadata, events",
+        },
+    )
+    if execution["status"] in ("queued", "created"):
+        raise ExecutionQueued(f"Execution #{counter} is queued. Waiting for it to start...")
+    if execution["status"] in ("stopped", "stopping (hard)", "stopping", "error", "complete"):
+        raise ExecutionFinished(f"Execution #{counter}: Status({execution['status']}). Cannot SSH into it.")
+    return execution
+
+
+def get_ssh_details_from_execution(execution: dict) -> Tuple[str, int]:
+    events_response = fetch_status_events(execution, first_n=100)
+    ssh_details = find_ssh_details(list(events_response))
+    if not ssh_details:
+        raise click.UsageError("No SSH details found")
+    return ssh_details
+
+
+def get_ssh_details_with_retry(
+    counter: int,
+) -> Tuple[str, int]:
+    """
+    Fetch SSH details for a given execution
+
+    :param counter: The execution counter.
+    :return: A tuple of (IP address, port) for the SSH connection.
+    """
+    echo_attempts = 5
+    while True:
+        try:
+            execution = _get_execution_from_counter(counter)
+            break
+        except Exception as e:
+            if isinstance(e, ExecutionQueued):
+                echo_attempts -= 1
+                if echo_attempts > 0:
+                    click.echo(str(e))
+                time.sleep(5)
+                continue
+            raise
```

Comment:
I would inline `_get_execution_from_counter()` here and avoid the extra exception classes. `count()` is from `itertools`.

```suggestion
    project = get_project(require=True)
    assert project

    for attempt in count():
        execution = project.get_execution_from_counter(
            counter=counter,
            params={"exclude": "metadata, events"},
        )
        if execution["status"] in ("queued", "created"):
            if attempt % 3 == 0:  # print on first attempt and thereafter every 3 attempts
                click.echo(f"Execution #{counter} is {execution['status']}. Waiting for it to start...")
            time.sleep(5)
            continue
        if execution["status"] != "started":
            raise CLIException(f"Execution #{counter} is {execution['status']}. Cannot SSH into it.")
```